File: app/src/main/java/com/sp/base/ui/APICallingActivity.java
package com.sp.base.ui;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import com.google.gson.Gson;
import com.sp.base.R;
import com.sp.base.countrypicker.Country;
import com.sp.base.countrypicker.CountryPickerCallbacks;
import com.sp.base.countrypicker.CountryPickerDialog;
import com.sp.base.databinding.ActivityMainBinding;
import com.sp.base.network.ApiResponse;
import com.sp.base.ui.model.DummyResponseModel;
import com.sp.base.ui.model.MainViewModel;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class APICallingActivity extends AppCompatActivity {
    ActivityMainBinding mBinding;
    MainViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        observeData();
        callAPI();
    }

    public void getCountryPicker(Context mContext) {
        CountryPickerDialog countryPicker =
                new CountryPickerDialog(mContext, new CountryPickerCallbacks() {
                    @Override
                    public void onCountrySelected(Country country, int flagResId) {
                        Log.e("selected date", "onCountrySelected: " + country.getDialingCode());
                    }
                });
        countryPicker.show();
    }

    private void callAPI() {
        viewModel.getDataList();
    }

    private void observeData() {
        viewModel = new ViewModelProvider(this).get(MainViewModel.class);
        viewModel.getResponse().observe(this, new androidx.lifecycle.Observer<ApiResponse>() {
            @Override
            public void onChanged(ApiResponse apiResponse) {
                consumeResponse(apiResponse, "list");
            }
        });
    }


    private void consumeResponse(ApiResponse apiResponse, String tag) {
        switch (apiResponse.status) {
            case LOADING:
//                mBinding.progressIndicator.show();
                mBinding.result.setText("");

                break;

            case SUCCESS:
//                mBinding.progressIndicator.hide();
                if (!apiResponse.data.isJsonNull()) {
                    Log.e("response:", apiResponse.data.toString());
                    if (tag.equalsIgnoreCase("list")) {
                        DummyResponseModel dummyResponseModel = new Gson().fromJson(apiResponse.data.toString(), DummyResponseModel.class);
                        Log.e("response", "consumeResponse: " + dummyResponseModel.toString());
                        mBinding.result.setText("Result: " + dummyResponseModel.results.toString());
                    }
                }

                break;
            case ERROR:
                mBinding.result.setText("");
//                mBinding.progressIndicator.hide();
                break;
            default:
//                mBinding.progressIndicator.hide();
                mBinding.result.setText("");

                break;
        }
    }
}

File: app/src/main/java/com/baseapplication/ui/APICallingActivity.kt
package com.baseapplication.ui

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import androidx.paging.PagedList
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager.VERTICAL
import com.baseapplication.R
import com.baseapplication.databinding.ActivityMainBinding
import com.baseapplication.network.ApiResponse
import com.baseapplication.network.Status
import com.baseapplication.ui.adapter.ListDataAdapter
import com.google.gson.Gson
import com.onetuchservice.business.ui.model.DummyResponse
import com.onetuchservice.business.ui.model.MainViewModel
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class APICallingActivity() : AppCompatActivity() {
    private val TAG: String = APICallingActivity::class.java.getSimpleName()

    lateinit var mBinding: ActivityMainBinding
    lateinit var viewModel: MainViewModel
    var allNotificationListAdapter: ListDataAdapter? = null


    // PagedList configuration
    private val config = PagedList.Config.Builder()
        .setPageSize(20) // Page size (the number of items to load per page)
        .setEnablePlaceholders(false)
        .build()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        initData()
        observeData()
        viewModel?.getData(1, 20)

    }

    private fun initData() {
        val manager: LinearLayoutManager = LinearLayoutManager(this,VERTICAL,false)
        mBinding?.rv?.layoutManager = manager
        mBinding?.rv?.setHasFixedSize(true)
        allNotificationListAdapter = ListDataAdapter(this@APICallingActivity)
        mBinding?.rv?.adapter = allNotificationListAdapter
    }

    private fun observeData() {
        viewModel = ViewModelProvider(this).get(MainViewModel::class.java)
        viewModel?.responseLiveData?.observe(this) { apiResponse ->
            apiResponse?.let {
                consumeResponse(it, "list")
            }
        }
    }

    private fun consumeResponse(apiResponse: ApiResponse, tag: String) {

        Log.e("response:", apiResponse.data.toString())
        when (apiResponse.status) {
            Status.LOADING ->
                mBinding.progressIndicator.show();
            Status.SUCCESS -> {
                mBinding.progressIndicator.hide();
                if (apiResponse.data?.isJsonNull == false) {
                    if (tag.equals("list", ignoreCase = true)) {
                        val dummyResponse: DummyResponse = Gson().fromJson(apiResponse.data.toString(), DummyResponse::class.java)
//                        allNotificationListAdapter = AllNotificationListAdapter(this@APICallingActivity)
//                        mBinding?.rv?.adapter = allNotificationListAdapter
                        // DataSourceFactory from repository or data source

//                        val listData = LivePagedListBuilder(dummyResponse.results, config).build()
//                        allNotificationListAdapter?.submitList(dummyResponse.results)
                    }
                }
            }

            Status.ERROR -> Log.e(TAG, "consumeResponse error: " + apiResponse.error)

            else -> {
                Log.e(TAG, "consumeResponse else: $apiResponse")
            }
        }
    }

}